javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
+      s2 + "."
+    else
+      s2
+  }
+
+  private def formatNonNumericString(flags: Flags, width: Int, precision: Int,
+      str: String): Unit = {
+
+    val truncatedStr =
+      if (precision < 0) str
+      else str.substring(0, precision)
+    padAndSendToDest(flags, width, applyUpperCase(flags, truncatedStr))
+  }
+
+  private def formatNaNOrInfinite(flags: Flags, width: Int, x: Double): Unit = {
+    val str = if (x.isNaN) {
+      "NaN"
+    } else if (x > 0.0) {
+      if (flags.positivePlus) "+Infinity"
+      else if (flags.positiveSpace) " Infinity"
+      else "Infinity"
+    } else {
+      if (flags.negativeParen) "(Infinity)"
+      else "-Infinity"
+    }
+
+    padAndSendToDest(flags, width, applyUpperCase(flags, str))
+  }
+
+  private def formatNumericString(flags: Flags, width: Int, str: String): Unit = {
+
+    val TransformativeFlags =
+      PositivePlus | PositiveSpace | UseGroupingSeps | NegativeParen
+
+    if (str.length >= width && !flags.hasAnyOf(TransformativeFlags)) {
+      // Super-fast-path
+      sendToDest(applyUpperCase(flags, str))
+    } else if (!flags.hasAnyOf(TransformativeFlags | ZeroPad)) {
+      // Fast-path that does not need to inspect the string
+      formatNonNumericString(flags, width, -1, str)
+    } else {
+      // Extract prefix and rest, based on flags and the presence of a sign
+      val (prefix, rest0) = if (str.charAt(0) != '-') {
+        if (flags.positivePlus)
+          ("+", str)
+        else if (flags.positiveSpace)
+          (" ", str)
+        else
+          ("", str)
+      } else {
+        if (flags.negativeParen)
+          ("(", str.substring(1) + ")")
+        else
+          ("-", str.substring(1))
       }
+
+      // Insert grouping separators, if required
+      val rest =
+        if (flags.useGroupingSeps) insertGroupingSeps(rest0)
+        else rest0
+
+      // Apply uppercase, pad and send
+      padAndSendToDest(flags, width, prefix, applyUpperCase(flags, rest))
     }
+  }
 
-    this
+  private def insertGroupingSeps(s: String): String = {
+    val len = s.length
+    var index = 0
+    while (index != len && { val c = s.charAt(index); c >= '0' && c <= '9' }) {
+      index += 1
+    }
+
+    index -= 3
+
+    if (index <= 0) {
+      s
+    } else {
+      var result = s.substring(index)
+      while (index > 3) {
+        val next = index - 3
+        result = s.substring(next, index) + "," + result
+        index = next
+      }
+      s.substring(0, index) + "," + result
+    }
   }
 
-  def ioException(): IOException = null
-  def locale(): Locale = ifNotClosed { null }
-  def out(): Appendable = ifNotClosed { dest }
+  private def applyUpperCase(flags: Flags, str: String): String =
+    if (flags.upperCase) str.toUpperCase
+    else str
 
-  override def toString(): String = out().toString()
+  /** This method ignores `flags.zeroPad` and `flags.upperCase`. */
+  private def padAndSendToDest(flags: Flags, width: Int, str: String): Unit = {
```

Comment:
This is a bit magic, that one overload doesn't support `zeroPad` and the other one does. Could we at least `assert(!flags.zeroPad)`?

Reply:
We actually call it when `flags.zeroPad` is true, *intentionally* disregarding it. I have made that clearer by renaming this overload to `padAndSendToDestNoZeroPad`.